common/zeroconf/zeroconf.cpp

```diff
@@ -0,0 +1,130 @@
+/*
+	Aseba - an event-based framework for distributed robot control
+	Copyright (C) 2007--2016:
+		Stephane Magnenat <stephane at magnenat dot net>
+		(http://stephane.magnenat.net)
+		and other contributors, see authors.txt for details
+	
+	This program is free software: you can redistribute it and/or modify
+	it under the terms of the GNU Lesser General Public License as published
+	by the Free Software Foundation, version 3 of the License.
+	
+	This program is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU Lesser General Public License for more details.
+	
+	You should have received a copy of the GNU Lesser General Public License
+	along with this program. If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include "zeroconf.h"
+#include "../utils/utils.h"
+#include "../utils/FormatableString.h"
+#include <typeinfo>
+#include <iostream>
+#include <iomanip>
+#include <map>
+#include <dashel/dashel.h>
+
+using namespace std;
+
+namespace Aseba
+{
+	using namespace Dashel;
+
+	//! Register and announce a target described by a name and a port
+	void Zeroconf::registerPort(const string name, const unsigned port, const string txt)
+	{
+        ZeroconfService z;
+        uint16_t len = txt.size();
+        const char* record = txt.c_str();
+        DNSServiceErrorType err = DNSServiceRegister(&z.serviceref,
+                                                     kDNSServiceFlagsDefault,
+                                                     0, // default all interfaces
+                                                     name.c_str(),
+                                                     "_aseba._tcp",
+                                                     NULL, // use default domain
+                                                     NULL, // use this host name
+                                                     htons(port),
+                                                     len, // TXT length
+                                                     record, // TXT record
+                                                     cb_Register,
+                                                     &z);
+        if (err != kDNSServiceErr_NoError)
+            throw runtime_error(FormatableString("Zeroconf DNSServiceRegister: error %0").arg(err));
+        else
+        {
+            DNSServiceErrorType err = DNSServiceProcessResult(z.serviceref); // block until daemon replies
+            if (err != kDNSServiceErr_NoError)
+                throw runtime_error(FormatableString("Zeroconf DNSServiceProcessResult: error %0").arg(err));
```

Comment:
I would define a subclass of `std::runtime_error` as a public subclass of zeroconf to allow user code to catch this exception specifically.